kernel/kernel-api/src/main/java/io/delta/kernel/internal/rowtracking/RowTracking.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.rowtracking;
+
+import io.delta.kernel.data.Row;
+import io.delta.kernel.internal.DeltaErrors;
+import io.delta.kernel.internal.SnapshotImpl;
+import io.delta.kernel.internal.TableFeatures;
+import io.delta.kernel.internal.actions.*;
+import io.delta.kernel.utils.CloseableIterable;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.DataFileStatistics;
+import java.io.IOException;
+import java.util.List;
+
+/** A collection of helper methods for working with row tracking. */
+public class RowTracking {
+  private RowTracking() {
+    // Empty constructor to prevent instantiation of this class
+  }
+
+  private static final int ADD_FILE_ORDINAL = SingleAction.FULL_SCHEMA.indexOf("add");
+
+  /**
+   * Assigns base row IDs and default row commit versions to {@link AddFile} actions before
+   * committing. If row tracking is not supported by the given protocol, returns the input actions
+   * unchanged.
+   *
+   * <p>When needed, it sets each {@link AddFile} action’s base row ID based on the current high
+   * watermark and increments the watermark accordingly. If a default row commit version is missing,
+   * it assigns the provided commit version.
+   *
+   * @param protocol the protocol to check for row tracking support
+   * @param snapshot the current snapshot of the table
+   * @param commitVersion the version of the commit for default row commit version assignment
+   * @param dataActions the {@link CloseableIterable} of data actions to process
+   * @return an {@link CloseableIterable} of data actions with base row IDs and default row commit
+   *     versions assigned
+   */
+  public static CloseableIterable<Row> assignBaseRowIdAndDefaultRowCommitVersion(
```

Comment:
Another possible approach is to assign `baseRowId` and `defaultRowCommitVersion` when creating `AddFile` actions. 

However, this would duplicate logic across all places where `AddFile` actions are created (currently just `Transaction.generateAppendActions`, but potentially more in the future I guess). Plus, `baseRowId` assignment is stateful, making it tricky if `AddFile` actions are created in multiple steps/places.

I chose to handle it here for now to keep things centralized, even though it requires converting `AddFile` rows back to actions, updating them, and converting back to rows.